### ISSUE TITLE
Low accuracy for certain recognition models

### ISSUE THREAD

**Issue by KHTee** [serengil/deepface#293]
Hi, I try to check the accuracy of a few recognition models by using the test data provided [here ](https://github.com/serengil/deepface/tree/master/tests/dataset) and testing code in this [notebook](https://github.com/serengil/tensorflow-101/blob/master/python/LFW.ipynb). I'm using `euclidean_l2` as my distance score. Referring to issue #287, the accuracy should be around 99% for Euclidean and l2.
However, I'm only getting 96.66% for FaceNet while some of the models are having very low accuracy. Is there any problem in my implementation?

| Model | Accuracy | Precision | Recall | F1 |
| ------------- | ------------- | ------------- | ------------- | ------------- | 
| FaceNet | 96.66 % | 96.66% | 76.31 % | 85.29 % |
| Dlib | 24.33 % | 14.33 % | 100.0 % | 25.08 % |
| DeepFace | 79.0 % | 31.34 % | 55.26 % | 40.0 % |
| VGG-Face| 95.0 % | 100.0 % | 60.52 % | 75.40 % |

Below is my implementation
```
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from deepface import DeepFace
from tqdm import tqdm
import csv
from deepface.basemodels import VGGFace, Facenet, OpenFace, FbDeepFace
from deepface.basemodels.DlibResNet import DlibResNet

vgg_model = VGGFace.loadModel()
print("VGG-Face loaded")

facenet_model = Facenet.loadModel()
print("FaceNet loaded")

openface_model = OpenFace.loadModel()
print("OpenFace loaded")

deepface_model = FbDeepFace.loadModel()
print("DeepFace loaded")

dlib_model = DlibResNet()
print("Dlib loaded")

load_model = {
    'VGG-Face': vgg_model,
    'FaceNet': facenet_model,
    'OpenFace': openface_model,
    'DeepFace': deepface_model,
    'Dlib': dlib_model,
}

with open('tests/dataset/master.csv', newline='') as csvfile:
    csv_rows = csv.reader(csvfile, delimiter=',')
    csv_rows= list(csv_rows)[1:]

actuals = []; predictions = []; distances = []
model = 'Dlib'
pbar = tqdm(range(0, len(csv_rows)))

for i in pbar:
    img1 = 'tests/dataset/' + csv_rows[i][0]
    img2 = 'tests/dataset/' + csv_rows[i][1]

    obj = DeepFace.verify(img1, 
                          img2, 
                          model_name = model, 
                          model = load_model[model], 
                          distance_metric = 'euclidean_l2',
                          enforce_detection=False)

    prediction = obj["verified"]
    predictions.append(prediction)

    distances.append(obj["distance"])

    label = csv_rows[i][2]
    actual = True if label == "Yes" else False
    actuals.append(actual)

accuracy = 100*accuracy_score(actuals, predictions)
precision = 100*precision_score(actuals, predictions)
recall = 100*recall_score(actuals, predictions)
f1 = 100*f1_score(actuals, predictions)

print("instances: ",len(actuals))
print("accuracy: " , accuracy, "%")
print("precision: ", precision, "%")
print("recall: ", recall,"%")
print("f1: ",f1,"%")
```

My Python version is 3.6.13 and environment setting are 
```
Package                 Version
----------------------- -------------------
absl-py                 0.13.0
argon2-cffi             20.1.0
astunparse              1.6.3
async-generator         1.10
attrs                   21.2.0
backcall                0.2.0
bleach                  3.3.1
cached-property         1.5.2
cachetools              4.2.2
certifi                 2021.5.30
cffi                    1.14.6
charset-normalizer      2.0.3
click                   8.0.1
cmake                   3.21.0
cycler                  0.10.0
dataclasses             0.8
decorator               5.0.9
defusedxml              0.7.1
dlib                    19.22.0
entrypoints             0.3
filelock                3.0.12
Flask                   2.0.1
flatbuffers             1.12
gast                    0.3.3
gdown                   3.13.0
google-auth             1.33.1
google-auth-oauthlib    0.4.4
google-pasta            0.2.0
grpcio                  1.34.1
h5py                    2.10.0
idna                    3.2
importlib-metadata      4.6.1
ipykernel               5.5.5
ipython                 7.16.1
ipython-genutils        0.2.0
ipywidgets              7.6.3
itsdangerous            2.0.1
jedi                    0.18.0
Jinja2                  3.0.1
joblib                  1.0.1
jsonschema              3.2.0
jupyter                 1.0.0
jupyter-client          6.1.12
jupyter-console         6.4.0
jupyter-core            4.7.1
jupyterlab-pygments     0.1.2
jupyterlab-widgets      1.0.0
Keras                   2.4.3
keras-nightly           2.5.0.dev2021032900
Keras-Preprocessing     1.1.2
kiwisolver              1.3.1
lightgbm                3.2.1
Markdown                3.3.4
MarkupSafe              2.0.1
matplotlib              3.3.4
mistune                 0.8.4
mtcnn                   0.1.1
nbclient                0.5.3
nbconvert               6.0.7
nbformat                5.1.3
nest-asyncio            1.5.1
notebook                6.4.0
numpy                   1.19.5
oauthlib                3.1.1
opencv-contrib-python   4.5.3.56
opencv-python           4.5.3.56
opt-einsum              3.3.0
packaging               21.0
pandas                  1.1.5
pandocfilters           1.4.3
parso                   0.8.2
pexpect                 4.8.0
pickleshare             0.7.5
Pillow                  8.3.1
pip                     21.1.3
prometheus-client       0.11.0
prompt-toolkit          3.0.19
protobuf                3.17.3
ptyprocess              0.7.0
pyasn1                  0.4.8
pyasn1-modules          0.2.8
pycparser               2.20
Pygments                2.9.0
pyparsing               2.4.7
pyrsistent              0.18.0
PySocks                 1.7.1
python-dateutil         2.8.2
pytz                    2021.1
PyYAML                  5.4.1
pyzmq                   22.1.0
qtconsole               5.1.1
QtPy                    1.9.0
requests                2.26.0
requests-oauthlib       1.3.0
retina-face             0.0.4
rsa                     4.7.2
scikit-learn            0.24.2
scipy                   1.4.1
Send2Trash              1.7.1
setuptools              52.0.0.post20210125
six                     1.15.0
tensorboard             2.2.2
tensorboard-data-server 0.6.1
tensorboard-plugin-wit  1.8.0
tensorflow              2.2.0
tensorflow-estimator    2.2.0
termcolor               1.1.0
terminado               0.10.1
testpath                0.5.0
threadpoolctl           2.2.0
tornado                 6.1
tqdm                    4.61.2
traitlets               4.3.3
typing-extensions       3.7.4.3
urllib3                 1.26.6
wcwidth                 0.2.5
webencodings            0.5.1
Werkzeug                2.0.1
wheel                   0.36.2
widgetsnbextension      3.5.1
wrapt                   1.12.1
zipp                    3.5.0
```

**Comment by serengil** [comment]
1- Facenet, Facenet512, VGG-Face and ArcFace are better models. I don't prefer to use dlib normally.

2- I recommend you to use cosine or euclidean instead of l2

3- Setting mtcnn to detector_backend instead of opencv (default) will increase results.

My experiments are:

| Model | Accuracy | Precision | Recall | F1 |
| ------------- | ------------- | ------------- | ------------- | ------------- | 
| Dlib-Euclidean-OpenCV | 92 % | 64% | 84.21% | 72.72 % |
| Dlib-Cosine-OpenCV | 93.33 % | 69.56% | 84.21% | 76.19 % |
| Facenet512-Euclidean-OpenCV | 99 % | 94.87% | 97.36% | 96.10 % |
| Facenet512-Cosine-OpenCV | 97.66 % | 100% | 81.57% | 89.85 % |

**Comment by serengil** [comment]
I am updating the threshold values for model - distance pairs. I plan to push it tomorrow. That might solve your problem because I determine those threshold a long time ago and I changed detection & alignment procedures soon.

**Comment by serengil** [comment]
Just updated the thresholds. Clone the source code to have the latest one. I'm going to publish it in pip in the next release.
